CIPs/cip-137.md

```diff
@@ -0,0 +1,200 @@
+---
+cip: 137
+title: Ceramic API
+author: Danny Browning <dbrowning@3box.io>, Nathaniel Cook <nathaniel@3box.io>, Aaron Goldman <aaron@3box.io>, Joel Thorstensson <joel@3box.io>
+discussions-to: <URL to a topic under the CIP category on the Ceramic forum: https://forum.ceramic.network/c/cips>
+status: Draft
+category: Interface
+created: 2023-05-28
+edited: 2023-05-28
+requires: 124
+---
+
+
+## Simple Summary
+<!--Provide a simplified and layman-accessible explanation of the CIP.-->
+An API for reading and writing events on Ceramic event streams.
+
+
+## Abstract
+<!--A short (~200 word) description of the technical issue being addressed.-->
+TODO
+
+
+## Motivation
+<!--Motivation is critical for CIPs that want to change the Ceramic protocol. It should clearly explain why the existing protocol specification is inadequate to address the problem that the CIP solves. CIP submissions without sufficient motivation may be rejected outright.-->
+TODO
+
+
+## Specification
+<!--The technical specification should describe the syntax and semantics of any new feature.-->
+
+This is a work in progress specification that introduces a JSON-RPC interface for interacting directly with event streams on a node.
+
+
+
+### `ceramic_subscribe`
+
+This method tells the Ceramic node to subscribe to a given subset of streams. A stream subset is defined as `ceramic://*?<separator-key>=<separator-value>`, where *separator-key* describes a field in the header of the InitEvent of a stream,  and *separator-value* is the value of this field. For example to subscribe to all MIDs of a particular Model you can use `ceramic://*?model=kjzl6hvfrbw6c82mkud4qs38zl4hd03ifoyg2ksvfjkhuxebfzh3ef89vwvtvrr`.
+
+Once subscribed the Ceramic node will use the Recon protocol to keep in sync with other nodes subscribed to this stream set.
+
+Active subscriptions need to be persisted. Or can be stateless somehow? (Ask Nathaniel)
+
+**Params:**
+
+- *ceramicUrl* <string> - the Ceramic URL to subscribe to
+- *options* <object> - options for the subscription
+    - *rangeNumber* <integer> - splits the stream set space into the given number of ranges
+    - *rangeIndex* <integer> - subscribes only to the given range index
+
+**Returns:**
+
+- *range* <Array<string>> - a Recon message, representing the latest state of the range described by the subscription request. In the form of `[eventid, ahash, eventid]`
+
+---
+
+### `ceramic_unsubscribe`
+
+This method tells the Ceramic node to unsubscribe from a given stream set.
```

Comment:
If there are multiple clients that want the same stream, do we want something like reference counting.

e.g.
If each subscription was given a UUID and we tracked the controller that asked for it a controller could delete their subscription without deleting the other subscriptions for the same range.

1) alice subscribes to `ceramic://*?model=kjzl6hvfrbw6c82mkud4qs38zl4hd03ifoyg2ksvfjkhuxebfzh3ef89vwvtvrr`
2) bob subscribes to `ceramic://*?model=kjzl6hvfrbw6c82mkud4qs38zl4hd03ifoyg2ksvfjkhuxebfzh3ef89vwvtvrr`
3) bob unsubscribes to `ceramic://*?model=kjzl6hvfrbw6c82mkud4qs38zl4hd03ifoyg2ksvfjkhuxebfzh3ef89vwvtvrr`

at the end of 3 should the node be subscribed to `kjzl6hvfrbw6c82mkud4qs38zl4hd03ifoyg2ksvfjkhuxebfzh3ef89vwvtvrr`?

having the controller and UUID would give the unsubscriber the option to delete only their unsubscription or all unsubscriptions for `kjzl6hvfrbw6c82mkud4qs38zl4hd03ifoyg2ksvfjkhuxebfzh3ef89vwvtvrr`.